src/context/cdlist.h

```diff
@@ -327,111 +134,184 @@ class CDList : public ContextObj {
     }
 
-    // Postfix operations on iterators: requires a Proxy object to
-    // hold the intermediate value for dereferencing
-    class Proxy {
-      const T* d_t;
+   private:
+    const std::vector<T>* d_list;
+    size_t d_pos;
+  }; /* class CDList<>::const_iterator */
+  typedef const_iterator iterator;
 
-    public:
+  /**
+   * Returns an iterator pointing to the first item in the list.
+   */
+  const_iterator begin() const { return const_iterator(&d_list, 0); }
 
-      Proxy(const T& p): d_t(&p) {}
+  /**
+   * Returns an iterator pointing one past the last item in the list.
+   */
+  const_iterator end() const { return const_iterator(&d_list, d_size); }
 
-      T& operator*() {
-        return *d_t;
-      }
-    };/* class CDList<>::const_iterator::Proxy */
+ protected:
+  /**
+   * Private copy constructor used only by save(). d_list is not copied: only
+   * the base class information and d_size are needed in restore.
+   */
+  CDList(const CDList& l)
+      : ContextObj(l),
+        d_size(l.d_size),
+        d_callCleanup(false),
+        d_cleanUp(l.d_cleanUp)
+  {
+    Debug("cdlist") << "copy ctor: " << this << " from " << &l << " size "
+                    << d_size << std::endl;
+  }
+  CDList& operator=(const CDList& l) CVC4_UNDEFINED;
+  /**
+   * Implementation of mandatory ContextObj method restore: simply
+   * restores the previous size.  Note that the list pointer and the
+   * allocated size are not changed.
+   */
 
-    /** Postfix increment: returns Proxy with the old value. */
-    Proxy operator++(int) {
-      Proxy e(*(*this));
-      ++(*this);
-      return e;
-    }
+  void restore(ContextObj* data) override
+  {
+    Debug("cdlist") << "restore " << this << " level "
+                    << this->getContext()->getLevel() << " data == " << data
+                    << " call dtor == " << this->d_callCleanup << std::endl;
+    truncateList(((CDList<T, CleanUp, Allocator>*)data)->d_size);
+    Debug("cdlist") << "restore " << this << " level "
+                    << this->getContext()->getLevel() << " size back to "
+                    << this->d_size << std::endl;
+  }
 
-    /** Postfix decrement: returns Proxy with the old value. */
-    Proxy operator--(int) {
-      Proxy e(*(*this));
-      --(*this);
-      return e;
+  /**
+   * Given a size parameter smaller than d_size, truncateList()
+   * removes the elements from the end of the list until d_size equals size.
+   *
+   * WARNING! You should only use this function when you know what you are
+   * doing. This is a primitive operation with strange context dependent
+   * behavior! It is up to the user of the function to ensure that the saved
+   * d_size values at lower context levels are less than or equal to size.
+   */
+  void truncateList(const size_t size)
+  {
+    Assert(size <= d_size);
+    if (d_callCleanup)
+    {
+      while (d_size != size)
+      {
+        --d_size;
+        d_cleanUp(&(d_list[d_size]));
```

Comment:
This is going to change the order in which the destructors are called on these objects. Though the clean-up order will be the same as the previous destruction order.

How did you check all of the instances that have a destructor/cleanup function for equivalence?